### PR TITLE
push to stable

### PR DESCRIPTION
**Pull request by vanhauser-thc**
This is the first push to stable for 3.0
There were a lot of changes with afl-cc how afl-fuzz works, etc and I tried to document these changes visibly in our main README.md at the top.
I expect this will flush up several issues we still might have, which is important before we do an official release.
I would propose aiming for a 1st December to 11th January for the release so that there is enough testing in between.

Not sure how much sense it makes to review this PR as it is ... huuuuuge. but well. best effort. :)

### CODE REVIEW

File: src/afl-fuzz-extras.c
Comment (domenukk) [inline src/afl-fuzz-extras.c]: Note that `afl_realloc` will never shrink the buffer size. This call has no effect.
(The idea behind `afl_realloc` is, that we reuse buffers as much as possible only reallocate if we really need to -> if the new buffer is larger than what we saw to this day).
Even if extras could grow in this function, the `!=` check would be redundant as it's done in realloc, too.

File: src/afl-fuzz-extras.c
Comment (domenukk) [inline src/afl-fuzz-extras.c]: Maybe add a "... skipping" note?

File: src/afl-fuzz-init.c
Comment (domenukk) [inline src/afl-fuzz-init.c:716]: Did we ever check if we have issues with sharedmap input, mutations, splicing, ... when testcases are `MAX_FILE` sized?

File: src/afl-fuzz-init.c
Comment (domenukk) [inline src/afl-fuzz-init.c:1048]: so many gotos... hard to follow and may be problematic for the compiler

**Review by domenukk — APPROVED** [review]
Oh my...
Quite a lot of code :)
Seems to work, so let's do it?

File: src/afl-fuzz-extras.c
Comment (vanhauser-thc) [reply src/afl-fuzz-extras.c]: uh true. so what is the correct way?

File: src/afl-fuzz-extras.c
Comment (vanhauser-thc) [reply src/afl-fuzz-extras.c]: done

File: src/afl-fuzz-init.c
Comment (vanhauser-thc) [reply src/afl-fuzz-init.c:1048]: two gotos, and IMHO easy to read. also works like a charm. not using goto will make this much more complex.

File: src/afl-fuzz-init.c
Comment (vanhauser-thc) [reply src/afl-fuzz-init.c:716]: me not explicitly. I know some fuzzbench inputs are larger than that.
but this needs a target that actually uses that amount - otherwise it will be trimmed ...
we could test with perl/php/json targets or pcap

File: src/afl-fuzz-extras.c
Comment (domenukk) [reply src/afl-fuzz-extras.c]: I would just not resize it here, we may add extras later and need to re-resize, right? Alternatively, switch to normal malloc. I can also add an `afl_realloc_shrinking` if you want?

Comment (vanhauser-thc) [reply src/afl-fuzz-extras.c]: or afl_realloc_exact ... yes I think that would be a good thing to have

File: docs/FAQ.md
Comment (hexcoder-) [inline docs/FAQ.md]: I don't understand why all of my editing was thrown away. Merge error?

Comment (vanhauser-thc) [reply docs/FAQ.md]: me neither ... dunno why that happened

File: src/afl-fuzz-extras.c
Comment (vanhauser-thc) [reply src/afl-fuzz-extras.c]: -> a2739ef ?

Comment (vanhauser-thc) [reply src/afl-fuzz-extras.c]: plus ...  b33306c

File: docs/FAQ.md
Comment (hexcoder-) [reply docs/FAQ.md]: Can you resurrect my editing, please?

Comment (vanhauser-thc) [reply docs/FAQ.md]: no time, sorry, two customer projects

**Comment by vanhauser-thc** [comment]
IMHO @andreafioraldi needs to document the new qemu environment variables before we can merge into stable

File: docs/FAQ.md
Comment (vanhauser-thc) [reply docs/FAQ.md]: done

**Comment by vanhauser-thc** [comment]
dont merge yet, still testing if a fix is working

**Comment by vanhauser-thc** [comment]
and we should fix #562 ... @domenukk / @hexcoder-  ?

**Comment by hexcoder-** [comment]
> and we should fix #562 ... @domenukk / @hexcoder- ?

ok, i will have a look.

**Comment by vanhauser-thc** [comment]
@hexcoder- @domenukk @andreafioraldi IMHO we can merge now?

**Comment by andreafioraldi** [comment]
Still debugging the stability issue of the persistent hook. But IMO yes, it is a minor feature, when i'll figure out it we can do another PR

**Review by andreafioraldi — APPROVED** [review]

**Review by hexcoder- — APPROVED** [review]
Nice work!

**Comment by vanhauser-thc** [comment]
#597 - if this is really a bug on our side - affects both stable and dev (@andreafioraldi ), so this is not a blocker
#562 is IMHO fixed but I doubt 2 of the 3 reporters will retest
the other issues do not apply or are fixed IMHO.
so I am merging ...

**Pull request merged**
